### ISSUE TITLE
[no-unnecessary-condition] can't detect unnecessary a === a where a is object/array

### ISSUE THREAD

**Issue by AbdelrahmanHafez** [typescript-eslint/typescript-eslint#1735]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

```TS
const a = { hello: 'world' };

if (a === a) {
  console.log('Hello world');
}
```

**Expected Result**
Expected the linter to throw an error for the unnecessary if statement because `a === a` is always gonna be true except if `a` were NaN, and since we know it's an object, it _will_ be true.

**Actual Result**
No errors were thrown.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.22.0` |
| `@typescript-eslint/parser`        | `2.22.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.13.0` |
| `npm`                              | `6.12.0` |

**Comment by bradzacher** [comment]
This is actually really hard to implement. Typescript does not differentiate between the individual values, only the types.

In the example of object types, that means that when we inspect an identifier, we see a type with an id, a symbol with an id, and that's it.

```ts
type T = { a: string }
const a: T = { a: '' }
const b: T = { a: '' }
const c = a;

if (a === a) {}
if (a === b) {}
if (a === c) {}
```


So it is possible to detect `a === a`, because we can compare the symbols and see they are the same.
It is not possible to detect `a === c`, because even though it does always return true, from a typescript data point of view, `a === c` is the same as `a === b`.

You can see what the output looks like here:
https://ts-ast-viewer.com/#code/C4TwDgpgBAKlC8UDeUCGAuKBnYAnAlgHYDmUAvgFADGA9oTmpnIihlAOTvnV0MBGTBMkYculWvWBQqQ1AG4KFfADMoAClQJ4iVAEpklFes3bEffUkOqNWxFQuUgA

Happy to accept a PR if you want the rule to handle `a === a`, but the other cases aren't possible without complex control-flow analysis.

**Comment by ark120202** [comment]
`a === a` is already handled by [`no-self-compare`](https://eslint.org/docs/rules/no-self-compare). It doesn't do

> except if `a` were NaN

but using it to detect NaN is a bad practice anyway

**Comment by bradzacher** [comment]
So based on the last part already being a rule already, and the rest of it being pretty close to impossible to implement, I'm going to close this.
